types/wlr_layer_shell.c

```diff
@@ -0,0 +1,384 @@
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_layer_shell.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "wlr-layer-shell-unstable-v1-protocol.h"
+
+static void resource_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct zwlr_layer_shell_v1_interface layer_shell_impl;
+static struct zwlr_layer_surface_v1_interface layer_surface_impl;
+
+static struct wlr_layer_client *layer_client_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_layer_shell_v1_interface,
+		&layer_shell_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_layer_surface *layer_surface_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_layer_surface_v1_interface,
+		&layer_surface_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void layer_surface_configure_destroy(
+		struct wlr_layer_surface_configure *configure) {
+	if (configure == NULL) {
+		return;
+	}
+	wl_list_remove(&configure->link);
+	free(configure);
+}
+
+static void layer_surface_handle_ack_configure(struct wl_client *client,
+		struct wl_resource *resource, uint32_t serial) {
+	struct wlr_layer_surface *surface = layer_surface_from_resource(resource);
+
+	bool found = false;
+	struct wlr_layer_surface_configure *configure, *tmp;
+	wl_list_for_each_safe(configure, tmp, &surface->configure_list, link) {
+		if (configure->serial < serial) {
+			layer_surface_configure_destroy(configure);
+		} else if (configure->serial == serial) {
```

Comment:
It looks to me like the way this is written, the comparison will happen for exactly one configure serial when it should consider all of them in the list.

Reply:
That's how it should be done. The client is free to ignore all configure requests but the last one.

Reply:
What if they are ack_configuring not the last one, but one further diwn in the list? They will get the error then.

Reply:
Why? The configure will be in the list, `found` will be set to `true`… All previous configures will be destroyed, but the next ones will remain, so the client will be able to ack_configure them in the future.